python/vote-text-bot/bot.py

```diff
@@ -0,0 +1,120 @@
+            'header': {'title': 'Last vote by %s!' % voter},
+            'sections': [{
+                'widgets': [{
+                    'textParagraph': {'text': '%d votes!' % vote_count}
+                }, {
+                    'buttons': [{
+                        'textButton': {
+                            'text': '+1',
+                            'onClick': {
+                                'action': {
+                                    'actionMethodName': 'upvote',
+                                    'parameters': PARAMETERS,
+                                }
+                            }
+                        }
+                    }, {
+                        'textButton': {
+                            'text': '-1',
+                            'onClick': {
+                                'action': {
+                                    'actionMethodName': 'downvote',
+                                    'parameters': PARAMETERS,
+                                }
+                            }
+                        }
+                    }, {
+                        'textButton': {
+                            'text': 'NEW',
+                            'onClick': {
+                                'action': {
+                                    'actionMethodName': 'newvote',
+                                }
+                            }
+                        }
+                    }]
+                }]
+            }]
+        }]
+    }
+
+class MainPage(webapp2.RequestHandler):
+    def post(self):
+        """Handle requests from Hangouts Chat
+
+        * Create new vote for ADDED_TO_SPACE and MESSAGE events
+        * Update existing card for 'upvote' or 'downvote' clicks
+        * Create new vote for 'newvote' clicks
+
+        @type  self:      RequestHandler()
+        @param self:      request/response object
+        """
+        event = json.loads(self.request.body)
+        user = event['user']['displayName']
+        # Update vote when card button pressed
+        if event['type'] == 'CARD_CLICKED':
+            method = event['action']['actionMethodName']
+            # Create new vote when "NEW" button clicked
+            if method == 'newvote':
+                body = createMessage(user)
+            # Update card in-place when "+1" or "-1" button clicked
+            else:
+                delta = 1 if method == 'upvote' else -1
+                vote_count = int(event['action']['parameters'][0]['value']) + delta
+                body = createMessage(user, vote_count, True)
+        # Create new vote for any msg
+        elif event['type'] == 'MESSAGE':
+            body = createMessage(user)
+        else: # no response for ADDED_TO_SPACE or REMOVED_FROM_SPACE
+            return
+
+        # Send response back to Hangouts Chat
+        self.response.headers['Content-Type'] = 'application/json'
+        self.response.out.write(json.dumps(body))
+
+# start server loop
```

Comment:
I don't think `start server loop` is completely accurate. I don't think there's explicit looping.

This snippet setups up a web server that listens to one path, `/`. Say that instead.

Reply:
DONE